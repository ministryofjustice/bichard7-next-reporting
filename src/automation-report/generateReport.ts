import { stringify } from "csv-stringify/sync"
import convertCsvToXlsx from "src/shared/csv-to-xlsx"
import { findForceName } from "src/shared/forces"
import type { AuditLog, AuditLogEvent, KeyValuePair } from "src/shared/types"

interface Force {
  exceptions: number
  manuallyResolved: number
  resubmittedAndResolved: number
  automated: number
  automatedQuotient: number
  automatedWithResubmissionQuotient: number
}

interface CalculateForceResult {
  force: Force
  national: Force
}

type Forces = KeyValuePair<string, Force>

const hasExceptions = (events: AuditLogEvent[]) =>
  events.some(
    (e) =>
      ["exceptions.generated", "exceptions.locked", "exceptions.unlocked", "exceptions.resolved"].includes(
        e.eventCode
      ) ||
      [
        "Exceptions generated",
        "Exception locked",
        "Exception unlocked",
        "Exception marked as resolved by user",
        "PNC Update marked as resolved"
      ].includes(e.eventType)
  )

const hasBeenResolvedManually = (events: AuditLogEvent[]) =>
  events.some(
    (e) =>
      e.eventCode === "exceptions.resolved" ||
      ["Exception marked as resolved by user", "PNC Update marked as resolved"].includes(e.eventType)
  )

const hasUpdatedPnc = (events: AuditLogEvent[]) =>
  events.some((e) => e.eventCode === "pnc.updated" || e.eventType === "PNC Update applied successfully") ||
  events.some(
    (e) =>
      (e.eventCode === "pnc.response-received" || e.eventType === "PNC Response received") &&
      e.attributes?.["PNC Request Type"] !== "ENQASI"
  )

const calculateForce = (events: AuditLogEvent[], force: Force, national: Force): CalculateForceResult => {
  const updatedForce = { ...force }
  const updatedNational = { ...national }

  const foundException = hasExceptions(events)
  if (foundException) {
    updatedForce.exceptions += 1
    updatedNational.exceptions += 1
  }

  if (hasBeenResolvedManually(events)) {
    updatedForce.manuallyResolved += 1
    updatedNational.manuallyResolved += 1
  }

  if (hasUpdatedPnc(events)) {
    if (foundException) {
      updatedForce.resubmittedAndResolved += 1
      updatedNational.resubmittedAndResolved += 1
    } else {
      updatedForce.automated += 1
      updatedNational.automated += 1
    }
  }

  return { force: updatedForce, national: updatedNational }
}

const calculateForces = (messages: AuditLog[]): Forces => {
  const forces: KeyValuePair<string, Force> = {}
  const newForce = () => ({
    exceptions: 0,
    manuallyResolved: 0,
    resubmittedAndResolved: 0,
    automated: 0,
    automatedQuotient: 0,
    automatedWithResubmissionQuotient: 0
  })

  let national = newForce()

  messages.forEach((message) => {
    const { forceOwner, events } = message
    const forceName = findForceName(forceOwner) || "Unknown"
    const force = (forceName && forces[forceName]) || newForce()
    const calculateForceResult = calculateForce(events, force, national)
    national = calculateForceResult.national
    if (forceName !== "Unknown") {
      forces[forceName] = calculateForceResult.force
    }
  })

  forces.national = national

  Object.keys(forces).forEach((forceName) => {
    const force = forces[forceName]
    const { automated, exceptions, manuallyResolved, resubmittedAndResolved } = force
    const automatedWithResubmissionDivider = automated + resubmittedAndResolved + exceptions - manuallyResolved

    if (automatedWithResubmissionDivider > 0) {
      force.automatedQuotient = (automated / automatedWithResubmissionDivider) * 100
      force.automatedWithResubmissionQuotient =
        ((automated + resubmittedAndResolved) * 100) / automatedWithResubmissionDivider
    }
  })

  return forces
}

const generateCsvLines = (date: Date, forces: Forces): string[][] => {
  const lines = [
    [`Bichard 7 Resubmissions - ${date.toDateString().split(" ")[1]} ${date.getUTCFullYear()}`],
    ["Force", "FullAutomation", "Number of Resubmissions", "Automated inc. Resubmissions"]
  ]
  const addLine = (forceName: string, name?: string) => {
    const force = forces[forceName]
    const automated = force.automatedQuotient.toFixed(2)
    const resubmittedAndResolved = force.resubmittedAndResolved.toFixed(0)
    const automatedWithResubmission = force.automatedWithResubmissionQuotient.toFixed(2)
    lines.push([`${name || forceName}`, `${automated}%`, `${resubmittedAndResolved}`, `${automatedWithResubmission}%`])
  }

  Object.keys(forces)
    .filter((forceName) => forceName !== "national")
    .sort()
    .forEach((forceName: string) => addLine(forceName))

  addLine("national", "National Average")

  return lines
}

export default (date: Date, events: AuditLog[]): Buffer => {
  const forces = calculateForces(events)

  const csvLinesResult = generateCsvLines(date, forces)

  const title = csvLinesResult.shift()?.[0]
  const xlsxResult = convertCsvToXlsx(stringify(csvLinesResult), title)

  return xlsxResult
}
