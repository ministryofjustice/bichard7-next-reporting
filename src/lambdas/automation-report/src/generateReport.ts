import { findForceName } from "@bichard/forces"
import type { AuditLog, AuditLogEvent, KeyValuePair } from "@bichard/types"
import { stringify } from "csv-stringify/sync"

interface Force {
  exceptions: number
  manuallyResolved: number
  resubmittedAndResolved: number
  automated: number
  automatedQuotient: number
  automatedWithResubmissionQuotient: number
}

interface CalculateForceResult {
  force: Force
  national: Force
}

type Forces = KeyValuePair<string, Force>

const exceptionsCategoryEventTypes = [
  "Hearing Outcome passed to Error List",
  "PNC Update added to Error List",
  "Exception marked as resolved by user"
]
const manuallyResolvedCategoryEventType = "Exception marked as resolved by user"
const pncUpdateSuccessfullyEventType = "PNC Update applied successfully"

const calculateForce = (events: AuditLogEvent[], force: Force, national: Force): CalculateForceResult => {
  const sortedEvents = events.sort((eventA, eventB) => (eventA.timestamp > eventB.timestamp ? 1 : -1))
  const updatedForce = { ...force }
  const updatedNational = { ...national }

  let foundException = false
  sortedEvents.forEach((event) => {
    const { eventType } = event

    if (exceptionsCategoryEventTypes.includes(eventType)) {
      updatedForce.exceptions += 1
      updatedNational.exceptions += 1
      foundException = true
    } else if (eventType === manuallyResolvedCategoryEventType) {
      updatedForce.manuallyResolved += 1
      updatedNational.manuallyResolved += 1
    } else if (eventType === pncUpdateSuccessfullyEventType) {
      updatedForce.automated += 1
      updatedNational.automated += 1

      if (foundException) {
        updatedForce.resubmittedAndResolved += 1
        updatedNational.resubmittedAndResolved += 1
      }
    }
  })

  return { force: updatedForce, national: updatedNational }
}

const calculateForces = (messages: AuditLog[]): Forces => {
  const forces: KeyValuePair<string, Force> = {}
  const newForce = () => ({
    exceptions: 0,
    manuallyResolved: 0,
    resubmittedAndResolved: 0,
    automated: 0,
    automatedQuotient: 0,
    automatedWithResubmissionQuotient: 0
  })

  let national = newForce()

  messages.forEach((message) => {
    const { forceOwner, events } = message.automationReport
    const forceName = findForceName(forceOwner)

    if (!forceName) {
      return
    }

    const force = forces[forceName] || newForce()

    const calculateForceResult = calculateForce(events, force, national)
    national = calculateForceResult.national
    forces[forceName] = calculateForceResult.force
  })

  forces.national = national

  Object.keys(forces).forEach((forceName) => {
    const force = forces[forceName]
    const { automated, exceptions, manuallyResolved, resubmittedAndResolved } = force
    const automatedWithResubmissionDivider = automated + exceptions - manuallyResolved

    if (automatedWithResubmissionDivider > 0) {
      force.automatedQuotient = automated / automatedWithResubmissionDivider
      force.automatedWithResubmissionQuotient = (automated + resubmittedAndResolved) / automatedWithResubmissionDivider
    }
  })

  return forces
}

const generateCsv = (forces: Forces): string => {
  const lines = [["Force", "Automated", "Resubmitted", "Automated including Resubmissions"]]
  const addLine = (forceName: string, name?: string) => {
    const force = forces[forceName]
    const automated = force.automatedQuotient.toFixed(5)
    const resubmittedAndResolved = force.resubmittedAndResolved.toFixed(5)
    const automatedWithResubmission = force.automatedWithResubmissionQuotient.toFixed(5)
    lines.push([`${name || forceName}`,`${automated}`,`${resubmittedAndResolved}`,`${automatedWithResubmission}`])
  }

  Object.keys(forces)
    .filter((forceName) => forceName !== "national")
    .sort()
    .forEach((forceName: string) => addLine(forceName))

  addLine("national", "National Average")

  //return lines.join("\n")
  return stringify(lines)
}

export default (events: AuditLog[]): string => {
  const forces = calculateForces(events)

  const csvResult = generateCsv(forces)

  return csvResult
}
