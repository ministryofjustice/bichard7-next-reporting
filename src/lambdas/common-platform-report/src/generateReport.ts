import { stringify } from "csv-stringify/sync"
import type { ReportRecord } from "./getReportData"

export default (records: ReportRecord[]): string => {
  const report = [
    ["Received Date", "Internal Message ID", "External Correlation ID", "PTIURN", "Error Message"]
  ].concat(records.map((r) => [r.receivedDate, r.messageId, r.externalCorrelationId, r.ptiurn, r.error]))
  return stringify(report)
}
