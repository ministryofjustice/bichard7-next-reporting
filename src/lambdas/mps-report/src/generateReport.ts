import getCourtErrors from "./getCourtErrors"
import { headers } from "../config/mpsHeadingKeys"
import getOffenceCode from "./getOffenceCode"
import AnnotatedHearingOutcome from "./types/AnnotatedHearingOutcome"
import { PostgresGateway } from "@bichard/postgres-gateway"
import { isError } from "@bichard/types"
import { xml2js } from "xml-js"
import { getText } from "./types/XmlStringCell"
import getResultQualifierCodes from "./getResultQualifierCodes"
import getDurationAndAmount from "./getDurationAndAmount"
import OffenceDetails, { isMultiple } from "./types/OffenceDetails"
import OffenceResult from "./types/OffenceResult"
import HearingOutcomeCase from "./types/HearingOutcomeCase"
import AnnotatedPNCUpdateDataset from "./types/AnnotatedPNCUpdateDataset"

interface ReportRowResultQuery {
    court_date: string
    court_code: string
    force_code: string
    is_urgent: string
    defendant_name: string
    triggers: string
    error_report: string
}

function updateCommonHeaders(row: ReportRowResultQuery, hearingOutcomeCase: HearingOutcomeCase, offence:OffenceDetails) {
  return [
    row.court_date, // Hearing Date
    row.court_code, // Court
    row.force_code, // Force
    row.is_urgent, // Urgent
    row.defendant_name, // Defendant
    getText(hearingOutcomeCase.HearingDefendant.DefendantDetail?.BirthDate), // DoB
    getText(hearingOutcomeCase.HearingDefendant.ArrestSummonsNumber), // ASN
    getText(hearingOutcomeCase.PTIURN), // URN
    row.triggers, // Triggers
    row.error_report, // Errors
    getOffenceCode(offence.CriminalProsecutionReference), // Offence Code
    getText(offence.ActualOffenceStartDate.StartDate), // Offence Start Date
    getText(offence.LocationOfOffence), // Offence Location
  ]
}

export default async (gateway: PostgresGateway) => {
  const rows = await getCourtErrors(gateway)
  if (isError(rows)) {
    return rows
  }
  
  let result = []
  result.push(headers.join(","))
  for (let i = 0; i < rows.length; i = i + 1) {
    const annotatedMsg: string = rows[i].annotated_msg.replace(/br7:/g, "").replace(/ds:/g, "")
    let annotatedMsgObject = (xml2js(annotatedMsg, { compact: true }) as AnnotatedPNCUpdateDataset)?.AnnotatedPNCUpdateDataset?.PNCUpdateDataset
    if(!annotatedMsgObject) {
      annotatedMsgObject = (xml2js(annotatedMsg, { compact: true }) as AnnotatedHearingOutcome)
    }
    const offences = annotatedMsgObject.AnnotatedHearingOutcome.HearingOutcome.Case.HearingDefendant.Offence
    const hearingOutcomeCase = annotatedMsgObject.AnnotatedHearingOutcome.HearingOutcome.Case

    if (isMultiple<OffenceDetails>(offences)) {
      for (let o = 0; o < offences.length; o = o + 1) {
        const newRow = updateCommonHeaders(rows[i], hearingOutcomeCase, offences[o])

        const results = offences[o].Result
        if (isMultiple<OffenceResult>(results)) {
          for (let j = 0; j < 10; j = j + 1) {
            if (j >= results.length) {
              newRow.push("") // result code
              newRow.push("") // result duration/amount
              newRow.push("") // result text
              newRow.push("") // result qualifier
            } else {
              // j-th
              newRow.push(getText(results[j].CJSresultCode)) // result code
              newRow.push(getDurationAndAmount(results[j])) // result duration/amount
              newRow.push(getText(results[j].ResultVariableText)) // result text
              newRow.push(getResultQualifierCodes(results[j].ResultQualifierVariable)) // result qualifier
            }
          }
        } else {
          newRow.push(getText(results.CJSresultCode)) // result code
          newRow.push(getDurationAndAmount(results)) // result duration/amount
          newRow.push(getText(results.ResultVariableText)) // result text
          newRow.push(getResultQualifierCodes(results.ResultQualifierVariable)) // result qualifier

          // fill in the remaining 9 groups of cells
          for (let j = 0; j < 9; j = j + 1) {
            newRow.push("") // result code
            newRow.push("") // result duration/amount
            newRow.push("") // result text
            newRow.push("") // result qualifier
          }
        }
        newRow.push(rows[i].trigger_locked_by_id) // Trigger Locked By
        newRow.push(rows[i].error_locked_by_id) // Exception Locked By

        result.push(newRow.join(","))
      }
    } else {
      const newRow = updateCommonHeaders(rows[i], hearingOutcomeCase, offences)

      const results = offences.Result
      if (isMultiple<OffenceResult>(results)) {
        for (let j = 0; j < 10; j = j + 1) {
          if (j >= results.length) {
            newRow.push("") // result code
            newRow.push("") // result duration/amount
            newRow.push("") // result text
            newRow.push("") // result qualifier
          } else {
            // j-th
            newRow.push(getText(results[j].CJSresultCode)) // result code
            newRow.push(getDurationAndAmount(results[j])) // result duration/amount
            newRow.push(getText(results[j].ResultVariableText)) // result text
            newRow.push(getResultQualifierCodes(results[j].ResultQualifierVariable)) // result qualifier
          }
        }
      } else {
        newRow.push(getText(results.CJSresultCode)) // result code
        newRow.push(getDurationAndAmount(results)) // result duration/amount
        newRow.push(getText(results.ResultVariableText)) // result text
        newRow.push(getResultQualifierCodes(results.ResultQualifierVariable)) // result qualifier

        // fill in the remaining 9 groups of cells
        for (let j = 0; j < 9; j = j + 1) {
          newRow.push("") // result code
          newRow.push("") // result duration/amount
          newRow.push("") // result text
          newRow.push("") // result qualifier
        }
      }
      newRow.push(rows[i].trigger_locked_by_id) // Trigger Locked By
      newRow.push(rows[i].error_locked_by_id) // Exception Locked By

      result.push(newRow.join(","))
    }
  }

  return result.join("\r\n")
}
