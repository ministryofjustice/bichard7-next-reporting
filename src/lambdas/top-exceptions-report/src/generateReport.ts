import { findForceName } from "@bichard/forces"
import type { AuditLog, KeyValuePair } from "@bichard/types"
import getErrorName from "./errorNames/getErrorName"
import { stringify } from "csv-stringify/sync"

type ForceExceptions = KeyValuePair<string, number>
type ForcesExceptions = KeyValuePair<string, ForceExceptions>

const getAttributesWithErrors = (messages: AuditLog[]): KeyValuePair<string, unknown>[] => {
  return messages
    .filter((x) => x.events && x.events.some((e) =>'Error 1 Details' in e.attributes))
    .map((x) => {
      const events = x.events
      for(let i=0; i< events.length; i+=1) {
        if(x.automationReport?.forceOwner) {
          events[i].attributes["Force Owner"] = x.automationReport.forceOwner
        }
      }
      return events
    })
    .reduce((allEvents, topExceptionsReportEvents) => allEvents.concat(topExceptionsReportEvents), [])
    .map((event) => event.attributes)
}

const calculateForcesExceptions = (allAttributes: KeyValuePair<string, unknown>[]): ForcesExceptions => {
  return allAttributes.reduce((forces, attributes) => {
    const forceOwner = attributes["Force Owner"] as string
    const forceName = findForceName(forceOwner) ?? "National"
    const forceExceptions = (forces[forceName] || {}) as ForceExceptions

    Object.keys(attributes)
      .filter((attributeName) => attributeName.match(/Error.*Details/))
      .forEach((key) => {
        const exceptionCode = (attributes[key] as string).split("||")[0]
        forceExceptions[exceptionCode] = (forceExceptions[exceptionCode] || 0) + 1
      })

    // eslint-disable-next-line no-param-reassign
    forces[forceName] = forceExceptions
    return forces
  }, {}) as ForcesExceptions
}

const generateCsv = (forcesExceptions: ForcesExceptions): string => {
  let rows = [["Force" ,"Exception" ,"Error Text" ,"Count"]]
  Object.keys(forcesExceptions).forEach((forceName) => {
    const forceExceptions = forcesExceptions[forceName]
    Object.keys(forceExceptions).forEach((exceptionCode) => {
      const errorText = getErrorName(exceptionCode)
      rows.push([`${forceName}`,`${exceptionCode}`,`${errorText}`,`${forceExceptions[exceptionCode]}`])
    })
  })
  return stringify(rows)
}

export default (messages: AuditLog[]): string => {
  const attributesWithErrors = getAttributesWithErrors(messages)

  const forcesExceptions = calculateForcesExceptions(attributesWithErrors)

  const csvResult = generateCsv(forcesExceptions)

  return csvResult
}
