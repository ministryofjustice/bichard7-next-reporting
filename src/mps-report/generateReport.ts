import { stringify } from "csv-stringify/sync"
import type { PostgresGateway } from "src/shared/postgres-gateway"
import type { PromiseResult } from "src/shared/types"
import { isError } from "src/shared/types"
import { xml2js } from "xml-js"
import headers from "./config/mpsHeadingKeys"
import type { CourtError } from "./getCourtErrors"
import getCourtErrors from "./getCourtErrors"
import getDurationAndAmount from "./getDurationAndAmount"
import getOffenceCode from "./getOffenceCode"
import getResultQualifierCodes from "./getResultQualifierCodes"
import type AnnotatedHearingOutcome from "./types/AnnotatedHearingOutcome"
import type AnnotatedPNCUpdateDataset from "./types/AnnotatedPNCUpdateDataset"
import type HearingOutcomeCase from "./types/HearingOutcomeCase"
import type OffenceDetails from "./types/OffenceDetails"
import { isMultiple } from "./types/OffenceDetails"
import type OffenceResult from "./types/OffenceResult"
import { getText } from "./types/XmlStringCell"

function updateCommonHeaders(row: CourtError, hearingOutcomeCase: HearingOutcomeCase, offence: OffenceDetails) {
  return [
    new Date(row.court_date).toISOString().split("T")[0], // Hearing Date
    row.court_code, // Court
    "1", // row.force_code, // Hard coded Force
    row.is_urgent === 1 ? "Y" : "N", // Urgent
    row.defendant_name, // Defendant
    getText(hearingOutcomeCase.HearingDefendant.DefendantDetail?.BirthDate), // DoB
    getText(hearingOutcomeCase.HearingDefendant.ArrestSummonsNumber), // ASN
    getText(hearingOutcomeCase.PTIURN), // URN
    row.triggers, // Triggers
    row.error_status === 1
      ? row.error_report
          .split(", ")
          .map((x) => x.split("|")[0])
          .join(" ")
      : "none", // Errors
    getOffenceCode(offence.CriminalProsecutionReference), // Offence Code
    getText(offence.ActualOffenceStartDate.StartDate), // Offence Start Date
    getText(offence.LocationOfOffence) // Offence Location
  ]
}

export default async (gateway: PostgresGateway): PromiseResult<string> => {
  const rows = await getCourtErrors(gateway)
  if (isError(rows)) {
    return rows
  }

  const result = []
  const now = new Date()
  result.push([
    "MPS Data Extract",
    `${now.getUTCDate().toString().padStart(2, "0")}/${(now.getUTCMonth() + 1)
      .toString()
      .padStart(2, "0")}/${now.getFullYear()} ${now.toISOString().split("T")[1].split(".")[0]}`,
    "",
    "NOT PROTECTIVELY MARKED"
  ])
  result.push([""])
  result.push(headers)
  for (let i = 0; i < rows.length; i += 1) {
    const annotatedMsg: string = rows[i].annotated_msg.replace(/br7:/g, "").replace(/ds:/g, "")
    let annotatedMsgObject = (xml2js(annotatedMsg, { compact: true }) as AnnotatedPNCUpdateDataset)
      ?.AnnotatedPNCUpdateDataset?.PNCUpdateDataset
    if (!annotatedMsgObject) {
      annotatedMsgObject = xml2js(annotatedMsg, { compact: true }) as AnnotatedHearingOutcome
    }
    const offences = annotatedMsgObject?.AnnotatedHearingOutcome?.HearingOutcome?.Case?.HearingDefendant?.Offence
    const hearingOutcomeCase = annotatedMsgObject?.AnnotatedHearingOutcome?.HearingOutcome?.Case

    if (!offences) {
      continue
    }
    if (isMultiple<OffenceDetails>(offences)) {
      for (let o = 0; o < offences.length; o += 1) {
        const newRow = updateCommonHeaders(rows[i], hearingOutcomeCase, offences[o])

        const results = offences[o].Result
        if (isMultiple<OffenceResult>(results)) {
          for (let j = 0; j < 10; j += 1) {
            if (j >= results.length) {
              newRow.push("") // result code
              newRow.push("") // result duration/amount
              newRow.push("") // result text
              newRow.push("") // result qualifier
            } else {
              // j-th
              newRow.push(getText(results[j].CJSresultCode)) // result code
              newRow.push(getDurationAndAmount(results[j])) // result duration/amount
              newRow.push(getText(results[j].ResultVariableText)) // result text
              newRow.push(getResultQualifierCodes(results[j].ResultQualifierVariable)) // result qualifier
            }
          }
        } else {
          newRow.push(getText(results.CJSresultCode)) // result code
          newRow.push(getDurationAndAmount(results)) // result duration/amount
          newRow.push(getText(results.ResultVariableText)) // result text
          newRow.push(getResultQualifierCodes(results.ResultQualifierVariable)) // result qualifier

          // fill in the remaining 9 groups of cells
          for (let j = 0; j < 9; j += 1) {
            newRow.push("") // result code
            newRow.push("") // result duration/amount
            newRow.push("") // result text
            newRow.push("") // result qualifier
          }
        }
        newRow.push(rows[i].trigger_locked_by_id ? rows[i].trigger_locked_by_id : "") // Trigger Locked By
        newRow.push(rows[i].error_locked_by_id ? rows[i].error_locked_by_id : "") // Exception Locked By

        result.push(newRow)
      }
    } else {
      const newRow = updateCommonHeaders(rows[i], hearingOutcomeCase, offences)

      const results = offences.Result
      if (isMultiple<OffenceResult>(results)) {
        for (let j = 0; j < 10; j += 1) {
          if (j >= results.length) {
            newRow.push("") // result code
            newRow.push("") // result duration/amount
            newRow.push("") // result text
            newRow.push("") // result qualifier
          } else {
            // j-th
            newRow.push(getText(results[j].CJSresultCode)) // result code
            newRow.push(getDurationAndAmount(results[j])) // result duration/amount
            newRow.push(getText(results[j].ResultVariableText)) // result text
            newRow.push(getResultQualifierCodes(results[j].ResultQualifierVariable)) // result qualifier
          }
        }
      } else {
        newRow.push(getText(results.CJSresultCode)) // result code
        newRow.push(getDurationAndAmount(results)) // result duration/amount
        newRow.push(getText(results.ResultVariableText)) // result text
        newRow.push(getResultQualifierCodes(results.ResultQualifierVariable)) // result qualifier

        // fill in the remaining 9 groups of cells
        for (let j = 0; j < 9; j += 1) {
          newRow.push("") // result code
          newRow.push("") // result duration/amount
          newRow.push("") // result text
          newRow.push("") // result qualifier
        }
      }
      newRow.push(rows[i].trigger_locked_by_id ? rows[i].trigger_locked_by_id : "") // Trigger Locked By
      newRow.push(rows[i].error_locked_by_id ? rows[i].error_locked_by_id : "") // Exception Locked By

      result.push(newRow)
    }
  }

  return stringify(result)
}
