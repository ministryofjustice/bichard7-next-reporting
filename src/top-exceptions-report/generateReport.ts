import { stringify } from "csv-stringify/sync"
import convertCsvToXlsx from "src/shared/csv-to-xlsx"
import { findForceName } from "src/shared/forces"
import type { AuditLog, KeyValuePair } from "src/shared/types"
import getErrorName from "./errorNames/getErrorName"

type ForceExceptions = KeyValuePair<string, number>
type ForcesExceptions = KeyValuePair<string, ForceExceptions>

const getAttributesWithErrors = (messages: AuditLog[]): KeyValuePair<string, unknown>[] => {
  return messages
    .filter(
      (message) => message.events && message.events.some((e) => e.attributes && "Error 1 Details" in e.attributes)
    )
    .map((message) => {
      const { events, forceOwner } = message
      for (let i = 0; i < events.length; i += 1) {
        if (forceOwner !== undefined && events[i].attributes) {
          events[i].attributes["Force Owner"] = forceOwner
        }
      }
      return events
    })
    .reduce((allEvents, topExceptionsReportEvents) => allEvents.concat(topExceptionsReportEvents), [])
    .map((event) => event.attributes ?? {})
}

const calculateForcesExceptions = (allAttributes: KeyValuePair<string, unknown>[]): ForcesExceptions => {
  return allAttributes.reduce((forces: ForcesExceptions, attributes) => {
    const forceNumber = attributes["Force Owner"] as number
    const forceName = findForceName(forceNumber) ?? "National"
    const forceExceptions = forces[forceName] || {}

    Object.keys(attributes)
      .filter((attributeName) => attributeName.match(/Error.*Details/))
      .forEach((key) => {
        const exceptionCode = (attributes[key] as string).split("||")[0]
        forceExceptions[exceptionCode] = (forceExceptions[exceptionCode] || 0) + 1
      })

    forces[forceName] = forceExceptions
    return forces
  }, {}) as ForcesExceptions
}

const generateCsv = (forcesExceptions: ForcesExceptions): string => {
  const rows = [["Force", "Exception", "Error Text", "Count"]]
  Object.keys(forcesExceptions).forEach((forceName) => {
    const forceExceptions = forcesExceptions[forceName]
    Object.keys(forceExceptions).forEach((exceptionCode) => {
      const errorText = getErrorName(exceptionCode)
      rows.push([`${forceName}`, `${exceptionCode}`, `${errorText}`, `${forceExceptions[exceptionCode]}`])
    })
  })
  return stringify(rows)
}

export default (messages: AuditLog[]): Buffer => {
  const attributesWithErrors = getAttributesWithErrors(messages)

  const forcesExceptions = calculateForcesExceptions(attributesWithErrors)

  const csvResult = generateCsv(forcesExceptions)

  const xlsxResult = convertCsvToXlsx(csvResult)

  return xlsxResult
}
