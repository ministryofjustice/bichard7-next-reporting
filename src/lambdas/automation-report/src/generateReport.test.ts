import generateReport from "./generateReport"
import messages from "../test/dummyMessages"

it("should generate the correct report", () => {
  const report = generateReport(new Date("2022-01-04T15:23:43.567Z"), messages(new Date("2021-10-29T10:34:00.000Z")))

  expect(report.toString()).toMatchSnapshot()
})
