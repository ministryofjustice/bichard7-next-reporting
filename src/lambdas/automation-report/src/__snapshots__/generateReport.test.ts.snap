// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should generate the correct report 1`] = `
"Force	Automated	Resubmitted	Automated including Resubmissions
Essex	0.68182	2.00000	0.77273
London MET	0.33333	0.00000	0.33333
National Average	0.64000	2.00000	0.72000"
`;
